ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt

```diff
@@ -442,6 +458,60 @@ private sealed class RuleExpr<out T> {
         //endregion
     }
 
+    sealed class IpExpr<T> : RuleExpr<T>() {
+        data class Ip(val value: RuleExpr<String>) : FunctionExpr<Long>() {
+            override fun eval(env: JsonObject): Long {
+                val octets = value.castEval<String>(env).split('.').map { it.toUByte() }
+                if (octets.size != 4) {
+                    throw IllegalArgumentException("invalid IPv4 format")
+                }
+
+                return calculateIpv4Value(octets)
+            }
+        }
+
+        data class Cidr(val value: RuleExpr<String>) : FunctionExpr<Collection<Long>>() {
+            override fun eval(env: JsonObject): Collection<Long> {
+                val splittedCidr = value.castEval<String>(env).split('/')
+                val subnetMask = if (splittedCidr.size > 1) splittedCidr[1].toByte() else 32
+                val octets = splittedCidr[0].split('.')
+
+                if (octets.size != 4) {
+                    throw IllegalArgumentException("invalid IPv4 format")
+                }
+
+                val mask = 0xFFFFFFFF shl (32 - subnetMask)
+                val subnet = arrayOf(
+                    (mask ushr 24).toUByte(),
+                    (mask shr 16 and 0xff).toUByte(),
+                    (mask shr 8 and 0xff).toUByte(),
+                    (mask and 0xff).toUByte()
+                )
+
+                val min = mutableListOf<UByte>(0u, 0u, 0u, 0u)
+                val max = mutableListOf<UByte>(0u, 0u, 0u, 0u)
+                octets.forEachIndexed { i, value ->
+                    val castedValue = value.toUByte()
+                    min[i] = castedValue and subnet[i]
+                    max[i] = castedValue or (subnet[i].inv())
+                }
+
+                return delegateRangeToCollection(calculateIpv4Value(min)..calculateIpv4Value(max))
+            }
+        }
+
+        companion object {
+            fun calculateIpv4Value(octets: List<UByte>): Long {
```

Comment:
```suggestion
            fun sumOctets(octets: List<UByte>): Long {
```

Weak suggestion, but “calculate IPv4 value” is a bit ambiguous. 😄